cpp/src/text/replace.cu

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/strings/detail/utilities.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/utilities/error.hpp>
+#include <nvtext/detail/tokenize.hpp>
+#include <nvtext/tokenize.hpp>
+#include <strings/utilities.cuh>
+#include <text/utilities/tokenize_ops.cuh>
+
+namespace nvtext {
+namespace detail {
+namespace {
+
+/**
+ * @brief Functor to replace tokens in each string.
+ *
+ * This tokenizes a string using the given d_delimiter and replaces any tokens that match
+ * a string in d_targets with those from the d_repls column. Strings with no matching tokens
+ * are left unchanged.
+ *
+ * This should be called first to compute the size of each output string and then a second
+ * time to fill in the allocated output buffer for each string.
+ */
+struct replace_tokens_fn {
+  cudf::column_device_view const d_strings;  ///< strings to tokenize
+  cudf::column_device_view const d_targets;  ///< strings to replace
+  cudf::column_device_view const d_repls;    ///< replacement strings
+  cudf::string_view const d_delimiter;       ///< delimiter characters for tokenizing
+  const int32_t* d_offsets{};
+  char* d_chars{};
+
+  __device__ cudf::size_type operator()(cudf::size_type idx)
+  {
+    if (d_strings.is_null(idx)) return 0;
+    auto const d_str = d_strings.element<cudf::string_view>(idx);
+    auto nbytes      = d_str.size_bytes();
+    auto in_ptr      = d_str.data();
+    auto out_ptr     = d_chars ? d_chars + d_offsets[idx] : nullptr;
+    auto last_pos    = cudf::size_type{0};
+    characters_tokenizer tokenizer(d_str, d_delimiter);
+    while (tokenizer.next_token()) {
+      auto token_pos = tokenizer.token_byte_positions();
+      cudf::string_view token{d_str.data() + token_pos.first, (token_pos.second - token_pos.first)};
+      // check if the token matches any of the targets
+      for (auto tidx = 0; tidx < d_targets.size(); ++tidx) {
+        cudf::string_view d_target = d_targets.element<cudf::string_view>(tidx);
+        if (token.compare(d_target) == 0) {
```

Comment:
This seems like a `find_if`. Whenever I see a `for` loop with a `break` at the end it always bugs me. Are we able to use the an algorithm here (I feel like I might have asked the before in a different PR and you said `cudf::string_view`s make this not possible.

Reply:
You may have asked this before in another API. I justified this here because I needed the resulting index value to match up the `d_targets` with `d_repls` columns. But you are right, I should be able to make use of `thrust::find` here instead of the for-loop.